File: ios/Runner/WebRTCClient.swift
import WebRTC
import SocketIO

@available(iOS 13.0, *)
class WebRTCClient: NSObject {
    private var peerConnections: [Int: RTCPeerConnection] = [:]
    private var userKeyTOname:[Int:String] = [:]
    private var mysetTrack: [Int:[String:Bool]] = [:]
    private var myAudio = true
    private var myVideo = true
//    var factory: RTCPeerConnectionFactory?
    var delegate: RTCPeerConnectionDelegate?
    var encoderFactory: RTCDefaultVideoEncoderFactory?
    var decoderFactory: RTCDefaultVideoDecoderFactory?
    var remoteStream: [Int: RTCVideoTrack] = [:]
    var remoteView: [Int: WebRTCView] = [:]
    var localStream: RTCMediaStream? // NativeView에서 전달받은 localStream
    var socket: SocketIOClient!
    private var manager: SocketManager!
    
    // 유저키값 FL에서 넘겨받아 적용
    var userKey: Int?
    private var roomKey: Int?
    private var userName: String?
    // 커넥션 완료 유무 FL로 전송
    var onConnectionCompleted: (() -> Void)?
    private var methodChannel: FlutterMethodChannel?
    
    override init() {
        super.init()
    }
    func setMethodChannel(_ methodChannel: FlutterMethodChannel) {
        self.methodChannel = methodChannel
    }
    func createPeerConnection() -> RTCPeerConnection? {
        // 인코더, 디코더 생성
        encoderFactory = RTCDefaultVideoEncoderFactory()
        decoderFactory = RTCDefaultVideoDecoderFactory()
        // 지원코덱
        for codecInfo in RTCDefaultVideoEncoderFactory.supportedCodecs() {
            if codecInfo.name.elementsEqual("H264") {
                encoderFactory!.preferredCodec = codecInfo
                break
            }
        }
        // 인코더, 디코더를 사용해 PeerConnectionFactory 생성
        let factory = RTCPeerConnectionFactory(encoderFactory: encoderFactory, decoderFactory: decoderFactory)
        let iceServers = [RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"])]
        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: [:])
        let config = RTCConfiguration()
        config.iceServers = iceServers
        config.iceTransportPolicy = .all
        config.rtcpMuxPolicy = .negotiate
        config.continualGatheringPolicy = .gatherContinually
        config.bundlePolicy = .maxBundle
        let peerConnection = factory.peerConnection(with: config, constraints: constraints, delegate: self)
        
        // 비디오 및 오디오 트랙을 peerConnection에 추가
        if let localStream = localStream {
            for track in localStream.videoTracks {
                peerConnection!.add(track, streamIds: [localStream.streamId])
            }
            for track in localStream.audioTracks {
                peerConnection!.add(track, streamIds: [localStream.streamId])
            }
        }
        return peerConnection
    }
    
    func connectWebRTC(roomKey: Int, userKey: Int, userName: String) {
        self.userKey = userKey
        self.roomKey = roomKey
        self.userName = userName
        // Socket.IO 초기화 및 연결
        manager = SocketManager(socketURL: URL(string: "http://192.168.0.9:3000")!, config: [.log(true), .compress, .forceNew(true), .forceWebsockets(true)])
        socket = manager.defaultSocket
        
        // 소켓 이벤트 소켓이 연결되면 서버에 join을 호출함
        socket.on(clientEvent: .connect) { data, ack in
            let joinInfo: [String: Any] = [
                "roomKey": self.roomKey!,
                "userKey": self.userKey!,
                "userName": self.userName!,
            ]
            self.socket.emit("join", joinInfo)
            self.onConnectionCompleted?()
        }
        
        socket.on("userJoin") { data, ack in
            guard let userData = data[0] as? [String: Any],
                  let userKey = userData["userKey"] as? Int,
                  let userName = userData["userName"] as? String else {
                return
            }
            
            if self.peerConnections[userKey] == nil{
                print("\(userKey) 방 들어옴")
                self.userKeyTOname[userKey] = userName
                self.mysetTrack[userKey] = ["oppAudioSet": true, "oppVideoSet": true, "myAudioSet": true , "myVideoSet": true]
                self.createOffer(userKey: userKey)}
        }
        socket.on("offer"){data, ack in
            guard let offerData = data[0] as? [String: Any],
                  let sdp = offerData["sdp"] as? String,
                  let userKey = offerData["userKey"] as? Int,
                  let userName = offerData["userName"] as? String else {
                return
            }
            
            if self.peerConnections[userKey] == nil{
                print("\(userKey) offer 들어옴")
                self.userKeyTOname[userKey] = userName
                self.mysetTrack[userKey] = ["oppAudioSet": true, "oppVideoSet": true, "myAudioSet": true , "myVideoSet": true]
                let remoteDescription = RTCSessionDescription(
                    type: RTCSdpType.offer, sdp: sdp)

                self.createAnswer(userKey: userKey, remoteDescription: remoteDescription)
            }
            
        }
        
        socket.on("answer") { data, ack in
            guard let answerData = data[0] as? [String: Any],
                  let sdp = answerData["sdp"] as? String,
                  let userKey = answerData["userKey"] as? Int,
                  let s_userKey = answerData["s_userKey"] as? Int else {
                return
            }
            if s_userKey == self.userKey! {
                print("\(userKey) answer들어옴")
                let sessionDescription = RTCSessionDescription(
                    type: RTCSdpType.answer, sdp: sdp)
                self.peerConnections[userKey]?.setRemoteDescription(sessionDescription) { error in
                    if error != nil {
                        return
                    }
                }
            }
            
        }
        
        socket.on("ice") { data, ack in
            guard let iceData = data[0] as? [String: Any],
                  let candidate = iceData["candidate"] as? String,
                  let sdpMid = iceData["sdpMid"] as? String,
                  let sdpMLineIndex = iceData["sdpMLineIndex"] as? Int32,
                  let userKey = iceData["userKey"] as? Int else {
                return
            }
            //            print("ice들어옴?")
            // ICE 연결 상태 확인
            if self.peerConnections[userKey]!.iceConnectionState == .connected || self.peerConnections[userKey]!.iceConnectionState == .completed {
                print("ICE 연결이 이미 완료되었습니다. ICE 후보를 추가하지 않습니다.")
                return
            }
            let iceCandidate = RTCIceCandidate(sdp: candidate, sdpMLineIndex: sdpMLineIndex, sdpMid: sdpMid)
            self.peerConnections[userKey]!.add(iceCandidate){ error in
                if error != nil {
                    print("아이스 추가 실패 \(userKey)")
                }else{
                    print("아이스 추가 완료 \(userKey)")
                }
            }
        }
        socket.on("audio"){ data, ack in
            guard let audioData = data[0] as? [String: Any],
                  let userKey = audioData["userKey"] as? Int,
                  let audio = audioData["audio"] as? Bool else {
                return
            }
            self.mysetTrack[userKey]!["oppAudioSet"] = audio
            self.audioSet(userKey:userKey)
            
        }
        socket.on("video"){ data, ack in
            guard let videoData = data[0] as? [String: Any],
                  let userKey = videoData["userKey"] as? Int,
                  let video = videoData["video"] as? Bool else {
                return
            }
            
            self.mysetTrack[userKey]!["oppVideoSet"] = video
            self.videoSet(userKey:userKey)
        }
        socket.on("exitUser"){ data, ack in
            guard let userKey = data[0] as? Int else {
                return
            }
            if self.peerConnections[userKey] != nil{
                print("\(userKey) 유저 나감")
                DispatchQueue.main.async {
                    self.methodChannel?.invokeMethod("exitUser", arguments: userKey)}
                if(self.remoteView[userKey] != nil){
                    self.remoteView[userKey]!.containerView.removeFromSuperview()
                }
                self.remoteView.removeValue(forKey: userKey)
                self.remoteStream.removeValue(forKey: userKey)
                self.peerConnections[userKey]!.close()
                self.peerConnections.removeValue(forKey: userKey)
                self.userKeyTOname.removeValue(forKey: userKey)
                self.mysetTrack.removeValue(forKey: userKey)
                
            }
        }
        self.socket.connect()
    }
    
    func createOffer(userKey: Int) {
        peerConnections[userKey] = createPeerConnection()
        let mandatory: [String: String] = [
            kRTCMediaConstraintsOfferToReceiveAudio: kRTCMediaConstraintsValueTrue,
            kRTCMediaConstraintsOfferToReceiveVideo: kRTCMediaConstraintsValueTrue
        ]
        let constraints = RTCMediaConstraints(mandatoryConstraints: mandatory, optionalConstraints: nil)
        peerConnections[userKey]!.offer(for: constraints) { (sdp, error) in
            guard let sdp = sdp else {
                return
            }
            self.peerConnections[userKey]!.setLocalDescription(sdp, completionHandler: { (error) in
                if error != nil {
                    return
                }
                // SDP 오퍼를 서버로 전송
                let offerData: [String: Any] = [
                    "sdp": sdp.sdp,
                    "userKey": self.userKey!,
                    "userName": self.userName!
                ]
                self.socket.emit("offer", offerData)
            })
        }
    }
    func createAnswer(userKey: Int, remoteDescription: RTCSessionDescription){
        peerConnections[userKey] = createPeerConnection()
        let mandatory: [String: String] = [
            kRTCMediaConstraintsOfferToReceiveAudio: kRTCMediaConstraintsValueTrue,
            kRTCMediaConstraintsOfferToReceiveVideo: kRTCMediaConstraintsValueTrue
        ]
        let constraints = RTCMediaConstraints(mandatoryConstraints: mandatory, optionalConstraints: nil)
        self.peerConnections[userKey]!.setRemoteDescription(remoteDescription) { error in
            if let error {
                print("Failed to set remote description: \(error)")
            }
        }
        self.peerConnections[userKey]!.answer(for: constraints) { (sdp, error) in
            guard let sdp = sdp else {
                return
            }
            self.peerConnections[userKey]!.setLocalDescription(sdp, completionHandler: { (error) in
                if error != nil {
                    return
                }
                // SDP answer를 서버로 전송
                let answerData: [String: Any] = [
                    "sdp": sdp.sdp,
                    "userKey": self.userKey!,
                    "s_userKey": userKey
                ]
                self.socket.emit("answer", answerData)
            })
        }
    }
    func audioChange(){
        self.myAudio = !self.myAudio
        let changeData: [String: Any] = [
            "userKey": self.userKey!,
            "audio": self.myAudio
        ]
        self.socket.emit("audio", changeData)
    }
    func videoChange(){
        self.myVideo = !self.myVideo
        let changeData: [String: Any] = [
            "userKey": self.userKey!,
            "video": self.myVideo
        ]
        self.socket.emit("video", changeData)
    }
    func oppAudioSet(userKey: Int){
        if let myAudioSet = self.mysetTrack[userKey]!["myAudioSet"] {
            self.mysetTrack[userKey]!["myAudioSet"] = !myAudioSet
        }
        audioSet(userKey: userKey)
    }
    func oppVideoSet(userKey: Int){
        if let myVideoSet = self.mysetTrack[userKey]!["myVideoSet"] {
            self.mysetTrack[userKey]!["myVideoSet"] = !myVideoSet
        }
        videoSet(userKey: userKey)
    }
    
    func audioSet(userKey: Int){
        if let myAudioSet = self.mysetTrack[userKey]!["myAudioSet"], myAudioSet {
            if let peerConnection = self.peerConnections[userKey] {
                for receiver in peerConnection.receivers {
                    if let track = receiver.track, track.kind == kRTCMediaStreamTrackKindAudio {
                        track.isEnabled = self.mysetTrack[userKey]!["oppAudioSet"]!
                    }
                }
            }
        } else {
            if let peerConnection = self.peerConnections[userKey] {
                for receiver in peerConnection.receivers {
                    if let track = receiver.track, track.kind == kRTCMediaStreamTrackKindAudio {
                        track.isEnabled = false
                    }
                }
            }
        }
    }
    func videoSet(userKey: Int){
        if let myVideoSet = self.mysetTrack[userKey]!["myVideoSet"], myVideoSet {
            if let peerConnection = self.peerConnections[userKey] {
                for receiver in peerConnection.receivers {
                    if let track = receiver.track, track.kind == kRTCMediaStreamTrackKindVideo {
                        track.isEnabled = self.mysetTrack[userKey]!["oppVideoSet"]!
                    }
                }
            }
        } else {
            if let peerConnection = self.peerConnections[userKey] {
                for receiver in peerConnection.receivers {
                    if let track = receiver.track, track.kind == kRTCMediaStreamTrackKindVideo {
                        track.isEnabled = false
                    }
                }
            }
        }
    }
    func exitUser(userKey: Int) {
        
    }
    
    func disconnectWebRTC() {
        for (_, peerConnection) in peerConnections {
            peerConnection.close()
        }
        peerConnections.removeAll()
        if(socket != nil && socket.status == .connected){
            socket.disconnect()
        }
    }
    func getUserKey(for peerConnection: RTCPeerConnection) -> Int? {
        for (key, value) in peerConnections {
            if value == peerConnection {
                return key
            }
        }
        return nil
    }
}

@available(iOS 13.0, *)
extension WebRTCClient: RTCPeerConnectionDelegate {
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange state: RTCSignalingState) {
    }
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
        
        if let userKey = getUserKey(for: peerConnection) {
            print("\(self.userKey!) 가 \(userKey) 의 스트림을 가져왔습니다.")
            self.remoteStream[userKey] = stream.videoTracks.first
            print("extenstion==== NTtoFL_remoteViewer call")
            let userData: [String:Any] = [
                "userKey": userKey,
                "userName": self.userKeyTOname[userKey]!
            ]
            let defaultAudio: [String: Any] = [
                "userKey": self.userKey!,
                "audio": self.myAudio
            ]
            let defaultVideo: [String: Any] = [
                "userKey": self.userKey!,
                "video": self.myVideo
            ]
            self.socket.emit("audio", defaultAudio)
            self.socket.emit("video", defaultVideo)
            DispatchQueue.main.async {
                self.methodChannel?.invokeMethod("NTtoFL_remoteViewer", arguments: userData)}
        }
    }
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {
    }
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange state: RTCIceConnectionState) {
    }
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange state: RTCIceGatheringState) {
    }
    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        let candidateData: [String:Any] = [
            "candidate": candidate.sdp,
            "sdpMid": candidate.sdpMid ?? "",
            "sdpMLineIndex": candidate.sdpMLineIndex,
            "userKey": self.userKey!,
        ]
        self.socket.emit("ice", candidateData);
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd receiver: RTCRtpReceiver, streams: [RTCMediaStream]) {
        
    }
}

File: android/app/src/main/kotlin/com/mooncorporation/rayo/WebRTCClient.kt
package com.soso.rayo


import android.content.Context
import android.util.Log
import android.view.ViewGroup
import io.flutter.plugin.common.MethodChannel
import io.socket.client.IO
import io.socket.client.Socket
import io.socket.engineio.client.transports.WebSocket
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.json.JSONObject
import org.webrtc.DataChannel
import org.webrtc.DefaultVideoDecoderFactory
import org.webrtc.DefaultVideoEncoderFactory
import org.webrtc.EglBase
import org.webrtc.IceCandidate
import org.webrtc.MediaConstraints
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.RtpReceiver
import org.webrtc.SdpObserver
import org.webrtc.SessionDescription
import org.webrtc.VideoTrack

class WebRTCClient(private val context: Context) {
    private var peerConnections: MutableMap<Int, PeerConnection> = mutableMapOf()
    private var userKeyTOname: MutableMap<Int, String> = mutableMapOf()
    private var mysetTrack: MutableMap<Int, MutableMap<String, Boolean>> = mutableMapOf()
    private var myAudio = true
    private var myVideo = true
//    private lateinit var factory: PeerConnectionFactory

    // TODO:: delegate 설정해야함
//    var delegate : PeerConnection??
    var encoderFactory: DefaultVideoEncoderFactory? = null
    var decoderFactory: DefaultVideoDecoderFactory? = null
    var remoteStream: MutableMap<Int, VideoTrack> = mutableMapOf()
    var remoteView: MutableMap<Int, WebRTCView> = mutableMapOf()
    var localStream: MediaStream? = null // NativeView에서 전달받은 localStream
    private lateinit var socket: Socket

    // 유저키값 FL에서 넘겨받아 적용
    var userKey: Int? = null
    private var roomKey: Int? = null
    private var userName: String? = null

    // 커넥션 완료 유무 FL로 전송
    var onConnectionCompleted: (() -> Unit)? = null
    private var methodChannel: MethodChannel? = null

    init {
        // WebRTC 초기화

    }

    fun setMethodChannel(methodChannel: MethodChannel) {
        this.methodChannel = methodChannel
    }

    @OptIn(DelicateCoroutinesApi::class)
    fun connectWebRTC(roomKey: Int, userKey: Int, userName: String) {
        this.roomKey = roomKey
        this.userKey = userKey
        this.userName = userName
        this.socket = IO.socket("http://192.168.0.9:3000", IO.Options().apply {
            transports = arrayOf(WebSocket.NAME)
            forceNew = true
            reconnection = true
        })
        this.socket.on(Socket.EVENT_CONNECT) {
            Log.d("NTLOG", "최초 접속 소켓 발송")
            val joinInfo = JSONObject().apply {
                put("roomKey", roomKey)
                put("userKey", userKey)
                put("userName", userName)
            }
            this.socket.emit("join", joinInfo)
            this.onConnectionCompleted?.let { it1 -> it1() }
        }
        this.socket.on("userJoin") { data ->
            val userData = data?.getOrNull(0) as? JSONObject ?: return@on
            val userKey = userData.get("userKey") as? Int ?: return@on
            val userName = userData.get("userName") as? String ?: return@on
            Log.d("NTLOG", "User Joined: $userKey, $userName")
            if (peerConnections[userKey] == null) {
                this.userKeyTOname[userKey] = userName
                this.mysetTrack[userKey] = mutableMapOf(
                    "oppAudioSet" to true,
                    "oppVideoSet" to true,
                    "myAudioSet" to true,
                    "myVideoSet" to true
                )
                this.createOffer(userKey)
            }
        }
        this.socket.on("offer") { data ->
            Log.d("NTLOG", "오퍼 요청받은 경우")
            val offerData = data?.getOrNull(0) as? JSONObject ?: return@on
            val sdp = offerData["sdp"] as? String ?: return@on
            val userKey = offerData.get("userKey") as? Int ?: return@on
            val userName = offerData.get("userName") as? String ?: return@on
            if (this.peerConnections[userKey] == null) {
                this.userKeyTOname[userKey] = userName
                this.mysetTrack[userKey] = mutableMapOf(
                    "oppAudioSet" to true,
                    "oppVideoSet" to true,
                    "myAudioSet" to true,
                    "myVideoSet" to true
                )
                val remoteDescription = SessionDescription(SessionDescription.Type.OFFER, sdp)
                this.createAnswer(userKey, remoteDescription)
                Log.d("NTLOG", "답변 완료 잘 된건가")
            }
        }

        this.socket.on("answer") { data ->
            Log.d("NTLOG", "오퍼에 대한 답변 요청의 경우")
            val answerData = data?.getOrNull(0) as? JSONObject ?: return@on
            val sdp = answerData["sdp"] as? String ?: return@on
            val userKey = answerData["userKey"] as? Int ?: return@on
            val s_userKey = answerData["s_userKey"] as? Int ?: return@on
            if (s_userKey == this.userKey) {
                val sessionDescription = SessionDescription(SessionDescription.Type.ANSWER, sdp)
                this.peerConnections[userKey]?.setRemoteDescription(
                    SimpleSdpObserver(),
                    sessionDescription
                )
            }

        }
        this.socket.on("ice") { data ->
            Log.d("NTLOG", "아이스 교환")
            val iceData = data?.getOrNull(0) as? JSONObject ?: return@on
            val candidate = iceData["candidate"] as? String ?: return@on
            val sdpMid = iceData["sdpMid"] as? String ?: return@on
            val sdpMLineIndex = iceData["sdpMLineIndex"] as? Int ?: return@on
            val userKey = iceData["userKey"] as? Int ?: return@on
            if (peerConnections[userKey] != null) {
                val iceConnectionState = peerConnections[userKey]?.iceConnectionState()
                if (iceConnectionState == PeerConnection.IceConnectionState.CONNECTED || iceConnectionState == PeerConnection.IceConnectionState.COMPLETED) {
                    return@on
                }
            }
            val iceCandidate = IceCandidate(sdpMid, sdpMLineIndex, candidate)
            this.peerConnections[userKey]?.addIceCandidate(iceCandidate)
        }
        this.socket.on("exitUser") { data ->
            val userKey = data?.getOrNull(0) as? Int ?: return@on

            if (this.peerConnections[userKey] != null) {
                Log.d("NTLOG", "$userKey 유저 나감")
                GlobalScope.launch(Dispatchers.Main) {
                    methodChannel?.invokeMethod("exitUser", userKey)
                }
                this.remoteView[userKey]?.let { webRTCView ->
                    (webRTCView.containerView.parent as? ViewGroup)?.removeView(webRTCView.containerView)
                }
//                this.remoteView[userKey]?.containerView?.removeView(this.remoteView[userKey])
                this.remoteStream.remove(userKey)
                this.peerConnections[userKey]!!.close()
                this.peerConnections.remove(userKey)
                this.userKeyTOname.remove(userKey)
                this.mysetTrack.remove(userKey)
            }
        };
        this.socket.connect()
    }

    private fun createOffer(userKey: Int) {
        Log.d("NTLOG", "오퍼 요청? 오퍼발송")
        this.peerConnections[userKey] = createPeerConnection(userKey)!!
        val mandatory = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "true"))
        }
        this.peerConnections[userKey]!!.createOffer(object : SimpleSdpObserver() {
            override fun onCreateSuccess(sessionDescription: SessionDescription) {
                this@WebRTCClient.peerConnections[userKey]!!.setLocalDescription(
                    SimpleSdpObserver(),
                    sessionDescription
                )
                val offerData = JSONObject().apply {
                    put("sdp", sessionDescription.description)
                    put("userKey", userKey)
                    put("userName", userName)
                }
                socket.emit("offer", offerData)
            }
        }, mandatory)

    }

    // answer
    private fun createAnswer(userKey: Int, remoteDescription: SessionDescription) {
        this.peerConnections[userKey] = createPeerConnection(userKey)!!
        val mandatory = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "true"))
        }
        this.peerConnections[userKey]!!.setRemoteDescription(SimpleSdpObserver(), remoteDescription)
        this.peerConnections[userKey]!!.createAnswer(object : SimpleSdpObserver() {
            override fun onCreateSuccess(sessionDescription: SessionDescription) {
                this@WebRTCClient.peerConnections[userKey]!!.setLocalDescription(
                    SimpleSdpObserver(),
                    sessionDescription
                )
                val answerData = JSONObject().apply {
                    put("sdp", sessionDescription.description)
                    put("userKey", this@WebRTCClient.userKey)
                    put("s_userKey", userKey)
                }
                socket.emit("answer", answerData)
            }

        }, mandatory)
    }


    fun createPeerConnection(userKey: Int): PeerConnection? {
        Log.d("NTLOG", "인코더 디코더 생성건")
        // 인코더, 디코더 생성
        encoderFactory = DefaultVideoEncoderFactory(
            EglBase.create().eglBaseContext, /* enableIntelVp8Encoder */
            true, /* enableH264HighProfile */
            true
        )
        decoderFactory = DefaultVideoDecoderFactory(EglBase.create().eglBaseContext)
        Log.d("NTLOG", "인코더 디코더 생성1")

        // 인코더, 디코더를 사용해 PeerConnectionFactory 생성
        val factory = PeerConnectionFactory.builder()
            .setVideoEncoderFactory(encoderFactory)
            .setVideoDecoderFactory(decoderFactory)
            .createPeerConnectionFactory()
        val iceServers = ArrayList<PeerConnection.IceServer>()
        val iceServerBuilder = PeerConnection.IceServer.builder("stun:stun.l.google.com:19302")
        iceServerBuilder.setTlsCertPolicy(PeerConnection.TlsCertPolicy.TLS_CERT_POLICY_INSECURE_NO_CHECK) //this does the magic.
        val iceServer = iceServerBuilder.createIceServer()
        iceServers.add(iceServer)
//        val iceServers = listOf(
//            PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer()
//        )
        Log.d("NTLOG", "인코더 디코더 생성2")
//        val constraints = MediaConstraints().apply {
//            optional.add(MediaConstraints.KeyValuePair("DtlsSrtpKeyAgreement", "true"))
//        }

        val config = PeerConnection.RTCConfiguration(iceServers)
        config.iceTransportsType = PeerConnection.IceTransportsType.ALL
        config.rtcpMuxPolicy = PeerConnection.RtcpMuxPolicy.NEGOTIATE
        config.continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY
        config.bundlePolicy = PeerConnection.BundlePolicy.MAXBUNDLE
        val peerConnection = factory.createPeerConnection(config, createPeerConnectionObserver(userKey))
        // 비디오 및 오디오 트랙을 peerConnection에 추가
        localStream?.let { stream ->
            stream.videoTracks.forEach { track ->
                peerConnection?.addTrack(track, listOf(stream.id))
            }
            stream.audioTracks.forEach { track ->
                peerConnection?.addTrack(track, listOf(stream.id))
            }
        }
        return peerConnection
    }

    fun disconnectWebRTC() {
        for ((_, peerConnection) in peerConnections) {
            peerConnection.close()
        }
        peerConnections.clear()
        if (this::socket.isInitialized && socket.connected()) {
            socket.disconnect()
        }
    }

    fun getUserKey(peerConnection: PeerConnection.Observer): Int? {
        for ((key, value) in peerConnections) {
            if (value == peerConnection) {
                return key
            }
        }
        return null
    }


    @OptIn(DelicateCoroutinesApi::class)
    private fun createPeerConnectionObserver(userKey: Int): PeerConnection.Observer {
        return object : PeerConnection.Observer {
            override fun onSignalingChange(signalingState: PeerConnection.SignalingState) {
                Log.d("NTLOG", "Signaling state changed: $signalingState")
            }

            override fun onIceConnectionChange(iceConnectionState: PeerConnection.IceConnectionState) {
                Log.d("NTLOG", "ICE connection state changed: $iceConnectionState")
            }

            override fun onIceConnectionReceivingChange(receiving: Boolean) {
                Log.d("NTLOG", "ICE connection receiving change: $receiving")
            }

            override fun onIceGatheringChange(iceGatheringState: PeerConnection.IceGatheringState) {
                Log.d("NTLOG", "ICE gathering state changed: $iceGatheringState")
            }

            override fun onIceCandidate(candidate: IceCandidate) {
                Log.d("NTLOG", "New ICE candidate: $candidate")
                val candidateData = JSONObject().apply {
                    put("candidate", candidate.sdp)
                    put("sdpMid", candidate.sdpMid)
                    put("sdpMLineIndex", candidate.sdpMLineIndex)
                    put("userKey", this@WebRTCClient.userKey)
                }
                socket.emit("ice", candidateData)
            }

            override fun onIceCandidatesRemoved(candidates: Array<IceCandidate>) {
                Log.d("NTLOG", "ICE candidates removed: ${candidates.joinToString()}")
            }

            override fun onAddStream(stream: MediaStream) {
                Log.d("NTLOG", "Stream added: $stream")
                remoteStream[userKey] = stream.videoTracks.first()
                val userData = mapOf(
                    "userKey" to userKey,
                    "userName" to userKeyTOname[userKey]!!
                )
                GlobalScope.launch(Dispatchers.Main) {
                    methodChannel?.invokeMethod("NTtoFL_remoteViewer", userData)
                }
            }

            override fun onRemoveStream(stream: MediaStream) {
                Log.d("NTLOG", "Stream removed: $stream")
            }

            override fun onDataChannel(dataChannel: DataChannel) {
                Log.d("NTLOG", "Data channel: $dataChannel")
            }

            override fun onRenegotiationNeeded() {
                Log.d("NTLOG", "Renegotiation needed")
            }

            override fun onAddTrack(receiver: RtpReceiver, streams: Array<MediaStream>) {
                Log.d("NTLOG", "Track added: $receiver")
            }
        }
    }

}

open class SimpleSdpObserver : SdpObserver {
    override fun onCreateSuccess(sessionDescription: SessionDescription) {
        Log.d("SimpleSdpObserver", "onCreateSuccess: $sessionDescription")
    }

    override fun onSetSuccess() {
        Log.d("SimpleSdpObserver", "onSetSuccess")
    }

    override fun onCreateFailure(error: String) {
        Log.e("SimpleSdpObserver", "onCreateFailure: $error")
    }

    override fun onSetFailure(error: String) {
        Log.e("SimpleSdpObserver", "onSetFailure: $error")
    }
}